.devcontainer/Dockerfile

```diff
@@ -0,0 +1,189 @@
+
+WORKDIR /uncrustify
+RUN curl -L https://github.com/uncrustify/uncrustify/archive/uncrustify-0.68.1.tar.gz | tar xz
+WORKDIR /uncrustify/uncrustify-uncrustify-0.68.1/
+RUN mkdir build
+WORKDIR /uncrustify/uncrustify-uncrustify-0.68.1/build/
+RUN cmake ..
+RUN make -sj8
+
+RUN make install DESTDIR=/uncrustify
+
+FROM base AS pipenv
+
+RUN sudo pip3 install pipenv pipenv-shebang
+
+WORKDIR /workspaces/citus/
+COPY src/ src/
+COPY .vscode/Pipfile .vscode/Pipfile.lock .devcontainer/.vscode/
+RUN ( cd src/test/regress && pipenv install )
+RUN ( cd .devcontainer/.vscode && pipenv install )
+
+# assemble the final container by copying over the artifacts from separately build containers
+FROM base AS devcontainer
+
+LABEL org.opencontainers.image.source=https://github.com/citusdata/citus
+LABEL org.opencontainers.image.description="Development container for the Citus project"
+LABEL org.opencontainers.image.licenses=AGPL-3.0-only
+
+RUN yes | sudo unminimize
+
+# install developer productivity tools
+RUN sudo apt update \
+ && sudo apt install -y \
+    autoconf2.69 \
+    bash-completion \
+    fswatch \
+    gdb \
+    htop \
+    libdbd-pg-perl \
+    libdbi-perl \
+    lsof \
+    man \
+    net-tools \
+    pspg \
+    vim \
+ && sudo apt clean
+
+RUN sudo pip3 install pipenv pipenv-shebang
+
+# add some common tools to the final container
+# bin directory for user tools
+RUN mkdir .bin
+ENV PATH="/home/citus/.bin:${PATH}"
+
+# install citus-dev
+RUN git clone --branch feature/citus-dev-pipenv-shebang https://github.com/citusdata/tools.git citus-tools \
+ && ( cd citus-tools/citus_dev && pipenv install ) \
+ && ln -s /home/citus/citus-tools/citus_dev/citus_dev-pipenv .bin/citus_dev \
+ && sudo make -C citus-tools/uncrustify install bindir=/usr/local/bin pkgsysconfdir=/usr/local/etc/
+
+# TODO some LC_ALL errors, possibly solved by locale-gen
+RUN git clone https://github.com/so-fancy/diff-so-fancy.git \
+ && ln -s /home/citus/diff-so-fancy/diff-so-fancy .bin/
+
+COPY --link --from=uncrustify-builder /uncrustify/usr/ /usr/
+
+COPY --link --from=pg14 /home/citus/.pgenv-staging/ /home/citus/.pgenv/
+COPY --link --from=pg15 /home/citus/.pgenv-staging/ /home/citus/.pgenv/
+COPY --link --from=pg16 /home/citus/.pgenv-staging/ /home/citus/.pgenv/
+
+COPY --link --from=pipenv /home/citus/.local/share/virtualenvs/ /home/citus/.local/share/virtualenvs/
+
+# place to run your cluster with citus_dev
+VOLUME /data
+RUN sudo mkdir /data \
+ && sudo chown citus:citus /data
+
+COPY --chown=citus:citus .psqlrc .
+
+# with the copy linking of layers github actions seem to misbehave with the ownership of the
+# directories leading upto the link, hence a small patch layer to have to right ownerships set
+RUN sudo chown --from=root:root citus:citus -R ~
+
+# sets default pg version
+RUN pgenv switch 16rc1
+
+USER root
+RUN curl -o /root/gdbpg.py https://raw.githubusercontent.com/tvesely/gdbpg/master/gdbpg.py
+COPY .gdbinit /root/
+USER citus
```

Comment:
We could probably move this up as it is very unlikely to change anymore

Also adding a comment on why we need to switch user to root here might give some guidance for a futuer reader, eg. myself